### PR TITLE
refactor(signing): support non-index zero input signing

### LINKED ISSUE #4620: Ledger wallet, sending an ordinal from regenerated address to primary taproot address fails with: The operation on device was rejected
When trying to send an ordinal from a regenerated address (produced by Ledger live) to another address via the flow in the wallet it fails with "The operation on device was rejected". 

image

image

### LINKED ISSUE #4628: Sending ordinals from regenerated addresses broke
Sending from the primary address to an other address is fine. I noticed an issue with Ledger and sending from regenerated addresses too (not sure if that is related: https://github.com/leather-wallet/extension/issues/4620)

Sending from a regenerated address to any other address is blocked at the point of selecting the fee... the wallet does not progress. 

Video about that:

https://github.com/leather-wallet/extension/assets/33360391/5e9e10bd-4508-454b-8897-f135802974e7

### PR DESCRIPTION
**Pull request by kyranjamie**
> Try out this version of Leather — [download extension builds](https://github.com/leather-wallet/extension/actions/runs/7141123809).

Closes #4620

Leather currently operates in a mode where addresses are reused. Users only have a single native segwit/taproot address per account. For a short while we were generating new addresses, per bitcoin privacy best practice, and built the signing logic it in a way to support this in future.

In refactoring the signing to work with Ledger via a unified API, I assumed the zero address index which broke this. This PR aims to fix this problem, by introducing a signing configuration, consisting of a derivation path and an input index. When creating the inscription transfer, we create this accompanying config, `BitcoinSigningConfig[]`. From this, we can then tell the signing function with which address index we need to sign the tx.

In order to not have to make changes in parts of the code where we do know transactions only need to be signed with the 0th index, I've written a function to make the config that assumes this.

Open to other ideas how we can do this. It's a shame in a way to have to have this accompanying metadata, but I can't think of any way we could otherwise know how to sign the tx at a later point in the tx signing flow.

### CODE REVIEW

**Comment by fbwoolf** [comment]
Related to this issue? https://github.com/leather-wallet/extension/issues/4645

**Comment by fbwoolf** [comment]
@kyranjamie I tested your branch. I still get the error and see the wrong address in their app, maybe it is on their end so will comment on the issue.

File: src/app/common/hooks/analytics/use-analytics.ts [hunk 1]
```diff
@@ -62,7 +62,7 @@ export function useAnalytics() {
         const [category, name, properties, options, ...rest] = args;
         const prop = { ...defaultProperties, ...properties };
         const opts = { ...defaultOptions, ...options };
-        logger.debug(`Analytics page view: ${name}`, properties);
+        // logger.debug(`Analytics page view: ${name}`, properties);
```
Comment (pete-watters) [inline src/app/common/hooks/analytics/use-analytics.ts]: Get rid of this maybe?

File: src/app/features/collectibles/components/bitcoin/ordinals.tsx [hunk 2]
```diff
@@ -19,6 +19,8 @@ export function Ordinals({ setIsLoadingMore }: OrdinalsProps) {
     rootMargin: '0% 0% 20% 0%',
   });
 
+  // console.log(query.data)
```
Comment (pete-watters) [inline src/app/features/collectibles/components/bitcoin/ordinals.tsx]: To delete

File: src/app/pages/psbt-request/use-psbt-request.tsx [hunk 3]
```diff
@@ -41,7 +43,10 @@ export function usePsbtRequest() {
         const tx = getPsbtAsTransaction(payload.hex);
 
         try {
-          const signedTx = await signPsbt({ indexesToSign: signAtIndex, tx });
+          const signedTx = await signPsbt({
```
Comment (pete-watters) [inline src/app/pages/psbt-request/use-psbt-request.tsx:46]: Using a `const` could make it a bit more readable here

```suggestion
         const signAtIndex = getDefaultSigningConfig(hexToBytes(payload.hex));
          const signedTx = await signPsbt({tx,
            indexesToSign: signAtIndex,
          });
```

File: src/app/common/hooks/analytics/use-analytics.ts (hunk 1 above)
Comment (pete-watters) [inline src/app/common/hooks/analytics/use-analytics.ts]:
```suggestion
```

File: src/app/common/hooks/analytics/use-analytics.ts [hunk 4]
```diff
@@ -73,7 +73,7 @@ export function useAnalytics() {
         const [eventName, properties, options, ...rest] = args;
         const prop = { ...defaultProperties, ...properties };
         const opts = { ...defaultOptions, ...options };
-        logger.debug(`Analytics event: ${eventName}`, properties);
+        // logger.debug(`Analytics event: ${eventName}`, properties);
```
Comment (pete-watters) [inline src/app/common/hooks/analytics/use-analytics.ts]:
```suggestion
```

File: src/app/features/collectibles/components/bitcoin/ordinals.tsx (hunk 2 above)
Comment (pete-watters) [inline src/app/features/collectibles/components/bitcoin/ordinals.tsx]:
```suggestion
```

File: src/app/pages/psbt-request/use-psbt-request.tsx [hunk 5]
```diff
@@ -41,7 +43,10 @@ export function usePsbtRequest() {
...
+            tx,
+            indexesToSign: getDefaultSigningConfig(hexToBytes(payload.hex)),
+          });
```
Comment (pete-watters) [inline src/app/pages/psbt-request/use-psbt-request.tsx]: Using a `const` could make it a bit more readable here

```suggestion
         const signAtIndex = getDefaultSigningConfig(hexToBytes(payload.hex));
          const signedTx = await signPsbt({tx,
            indexesToSign: signAtIndex,
          });
```

File: tests/mocks/constants.ts [hunk 6]
```diff
@@ -1,4 +1,6 @@
-// Bitcoin test addresses
+export const STANDARD_BIP_FAKE_MNEMONIC =
```
Comment (pete-watters) [inline tests/mocks/constants.ts:1]: Poor cactus 😢  🌵

File: tests/specs/rpc-sign-psbt/sign-psbt.spec.ts [hunk 7]
```diff
@@ -11,16 +11,18 @@ import { test } from '../../fixtures/fixtures';
...
+
 const signedOnlyIndexZeroPsbt =
   '70736274ff01007b02000000020c9199d8079e6fe8a6c78ac9c4e0311c97c9fcdc8b5586c56d191b6d98c0035e0000000000ffffffff087168f5b929b37a27704d338aa9d0d3508a819f879c244ba12128f04a5b37ef0000000000ffffffff01c800000000000000160014a8113965cee4d5ffa2d9996a204866a58200131d000000000001011f6400000000000000160014a8113965cee4d5ffa2d9996a204866a58200131d220203fe21e3444109e30ff7d19da0f530c344cad2e35fbee89afb2413858e4a9d7aa5483045022100ea4c2a68f1032102ad2c73504096f5dbd63d242ccce8000aa9db1a0ce4c4c59402204269fdd3536697329ed9bffcf67e3584d2d3426f84bb004fe467286abe7b02d8010001011f6400000000000000160014a8113965cee4d5ffa2d9996a204866a58200131d0000';
 
 test.describe('Sign PSBT', () => {
   test.beforeEach(async ({ extensionId, globalPage, onboardingPage, page }) => {
     await globalPage.setupAndUseApiCalls(extensionId);
     await onboardingPage.signInWithTestAccount(extensionId);
-    await page.goto('https://leather.io');
+    await page.goto('localhost:3000');
```
Comment (pete-watters) [inline tests/specs/rpc-sign-psbt/sign-psbt.spec.ts:25]: Just checking this is intentional?

**Review by pete-watters — APPROVED** [review]
Code LGTM  - didn't fully test the flow

File: src/shared/crypto/bitcoin/signer-config.spec.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import { getTaprootAccountDerivationPath } from './p2tr-address-gen';
+import { getNativeSegwitAccountDerivationPath } from './p2wpkh-address-gen';
+import { getAssumedZeroIndexSigningConfig } from './signer-config';
+
+describe(getAssumedZeroIndexSigningConfig.name, () => {
+  const seed = mnemonicToSeedSync(STANDARD_BIP_FAKE_MNEMONIC);
+  const keychain = HDKey.fromMasterSeed(seed);
+
+  // mock txid
+  const txid = '7fc33a83ba9627b8eeb0eebef90552f73518f27a45dbdc41bd6bd4342d098bf3';
```
Comment (fbwoolf) [inline src/shared/crypto/bitcoin/signer-config.spec.ts]: I think you could just name this `mockTxid`?

File: src/shared/crypto/bitcoin/signer-config.spec.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
+import { HDKey } from '@scure/bip32';
+import { mnemonicToSeedSync } from '@scure/bip39';
+import * as btc from '@scure/btc-signer';
```
Comment (fbwoolf) [inline src/shared/crypto/bitcoin/signer-config.spec.ts:3]: How is this working here? 🤔 Why do I run into issues with importing btc-signer in tests? Maybe now fixed with the playwright upgrade?

EDIT: I will go back and see if can refactor sign-psbt tests now with it.

**Review by fbwoolf — APPROVED** [review]
Nice refactor. 👍

File: tests/specs/rpc-sign-psbt/sign-psbt.spec.ts (hunk 7 above)
Comment (kyranjamie) [reply tests/specs/rpc-sign-psbt/sign-psbt.spec.ts:25]: Yep, cc @fbwoolf 

We just need a page to trigger the request from, better the local app than leather.io

File: src/app/common/hooks/analytics/use-analytics.ts (hunk 1 above)
Comment (kyranjamie) [reply src/app/common/hooks/analytics/use-analytics.ts]: https://github.com/pinojs/pino/issues/1861

File: src/shared/crypto/bitcoin/signer-config.spec.ts (hunk 9 above)
Comment (kyranjamie) [reply src/shared/crypto/bitcoin/signer-config.spec.ts:3]: These are unit tests, via Vitest, not playwright.

File: tests/specs/rpc-sign-psbt/sign-psbt.spec.ts (hunk 7 above)
Comment (fbwoolf) [reply tests/specs/rpc-sign-psbt/sign-psbt.spec.ts:25]: I took from other rpc signing tests, so best to change there too?

**Comment by kyranjamie** [comment]
@314159265359879 any chance you can QA pre-merge?

**Pull request merged**